### DISCUSSION TITLE
Migration of password hashes

### DISCUSSION THREAD

**Discussion by psachmann** [keycloak/keycloak#11794] [Q&A]
Hello,

we are currently in the process of migrating our custom user store including credentials to Keycloak. Our user store uses [BCrypt](https://en.wikipedia.org/wiki/Bcrypt) for password hashing and we want use Keycloak's default hashing algorithm. We observed that Keycloak will update old password hashes using the hashing algorithm set in the password policy. This allows a soft transition between BCrypt and Keycloak which is really handy for us. I couldn't find any documentation about this behavior and would like to ask, if this behavior is intended and can be used for a soft migration.

Kind regards,
Patrick

**Comment by MerlijnElderhuis** [comment]
Hi! I'm currently dealing with the same issue as you are. 
I found some resources about the way the data should be formatted, but logging still will not work.

Resources:
- https://github.com/keycloak/keycloak/blob/04415d34eab5ed577c472802df4ca11fbafabc7e/server-spi/src/main/java/org/keycloak/models/credential/dto/PasswordSecretData.java
- https://github.com/keycloak/keycloak/blob/04415d34eab5ed577c472802df4ca11fbafabc7e/server-spi/src/main/java/org/keycloak/models/credential/dto/PasswordCredentialData.java#L6
- CredentialRepresentation section on https://www.keycloak.org/docs-api/11.0/rest-api/#_credentialrepresentation
- and the discussion over here: https://groups.google.com/g/keycloak-user/c/XgUbXJr2xsM (even though they are not dealing with bcrypt over there)
- Also found this repo, which I think is aimed at including a bcrypt adapter within keycloak: https://github.com/leroyguillaume/keycloak-bcrypt

It seems that the data you need to send to the create_user endpoint should be in the following format, where `b64_salt` is actually the base64 encoded value of the salt part of the bcrypt password value.(we're using python so this is what I've tried)
```python
{
    "email": email,
    "username": email,
    "enabled": True,
    "emailVerified": True,
    "credentials": [
        {
            "type": "password",
            "credentialData": json.dumps({"hashIterations": 12, "algorithm": "bcrypt"}),
            "secretData": json.dumps({"value": value, "salt": b64_salt}),
        }
    ],
}
```
If anyone has some tips on how to continue, this would be very welcome. If I ever manage to do it, I will provide an update.

**Comment by MerlijnElderhuis** [comment]
Update:
Managed to get it working by installing a release of this repo in the keycloak application: https://github.com/leroyguillaume/keycloak-bcrypt

Used the curl command below (ran from inside a docker container, after which I restarted the container). 
ENV values:
- KEYCLOAK_BCRYPT_VERSION=1.5.0
- KEYCLOAK_HOME=~/keycloak

`curl -L https://github.com/leroyguillaume/keycloak-bcrypt/releases/download/${KEYCLOAK_BCRYPT_VERSION}/keycloak-bcrypt-${KEYCLOAK_BCRYPT_VERSION}.jar > ${KEYCLOAK_HOME}/standalone/deployments/keycloak-bcrypt-${KEYCLOAK_BCRYPT_VERSION}.jar`

After having done that, I could insert existing users in the db by using the following as input for the user_create endpoint:
`pw_hashed` is the hashed password including salt, so basically the old value in your db.
```python
{
    "email": email,
    "username": email,
    "enabled": True,
    "emailVerified": True,
    "credentials": [
        {
            "type": "password",
            "hashIterations": "12",
            "hashedSaltedValue": pw_hashed,
            "algorithm": "bcrypt",
        }
    ]
}
```

**Reply by psachmann** [reply]
Thank you for your answer. Maybe I was not precise enough. We have already a working proof of concept here <https://github.com/hpi-schul-cloud/erwin-idm/tree/feature/EW-133-bcrypt-spike>, which already uses [Keycloak BCrypt Plugin](https://github.com/leroyguillaume/keycloak-bcrypt) and [Keycloak Admin CLI](https://www.npmjs.com/package/@keycloak/keycloak-admin-client) to access Keycloak. Our main concern is how we migrate the password hashes, because we want to avoid having dependencies to third party plugins/libraries in the long run, that might need maintenance from us.

During the work on the proof we found out that Keycloak 17 + 18 (tested from us) will automatically update hashes to the default password hashing algorithm set in the password policy during a new user login. This behavior is not documented (or I didn't found it) and would greatly aid our migration process.
